### DISCUSSION TITLE
Can I use optional fields with the HTTP authenticator?

### DISCUSSION THREAD

**Discussion by agronholm** [emqx/emqx#9151] [Q&A] [answered]
I've tried to configure an HTTP based authenticator to accept either a username/password combo or an SSL client certificate. The problem is that if I try to add the `cert_subject` field to the body, then it fails with an interpolation error when there is no client certificate. I also tried the HOCON method of `${?cert_subject}` and `${?password}` but that does not seem to be supported, as those are passed as-is without any interpolation happening at all for those fields.

My next attempt was to configure two HTTP based authenticators, but for some reason EMQX does not seem to allow using the same back-end twice. Why?

So the question is: is it possible at all to configure authentication for username/password and client certificates with EMQX?

**Comment by hjianbo** [comment]
Thanks for your feedback, from your description you should be talking about the 5.0.x feature.

I like the solution you mentioned, using ${?password} as a placeholder for an optional field :thinking: 

However, I have a question. Why is there a cert_subject present in some requests and not in others? Could you briefly describe your usage scenario?

**Reply by agronholm** [reply]
End users are not issued client certificates, but we also have IoT gateways which would have those certificates through which they authenticate themselves to the broker. So end users authenticate using username/password, gateways authenticate using client certificates.

**Comment by lafirest** [comment, marked as answer]
@agronholm 
Hi, Which version are you using, I guess it might be v5?
In v5 we support using the `${cert_subject}` as the placeholder for the cert_subject, but there has a known issue: if you used this placeholder it must exist, we are trying to fix this.

Not being able to add the same type authenticator is a consideration for ease of use, maybe we could make a feature wanted issue

**Reply by agronholm** [reply]
> Hi, Which version are you using, I guess it might be v5?

Yes (5.0.8).

> In v5 we support using the ${cert_subject} as the placeholder for the cert_subject, but there has a known issue: if you used this placeholder it must exist, we are trying to fix this.

Yes, just having an empty string there instead of the interpolation failing completely would solve my issue. This is what I tried to achieve with `${?cert_subject}`.

**Reply by savonarola** [reply]
Hello!
>Yes, just having an empty string there instead of the interpolation failing

This is how the interpolation will work since 5.0.9

**Reply by agronholm** [reply]
Ah, fantastic! Thanks for letting me know.
